app/bundles/EmailBundle/Model/EmailModel.php

```diff
@@ -533,6 +542,99 @@ public function getBuilderComponents(Email $email = null, $requestedComponents =
+            if (!array_key_exists($statId, $data)) {
+                $item = [
+                    'contact_id'    => $stat['lead_id'],
+                    'contact_email' => $stat['email_address'],
+                    'open'          => $stat['is_read'],
+                    'email_id'      => $stat['email_id'],
+                    'email_name'    => $stat['email_name'],
+                ];
+                $item['click']         = ($stat['link_hits'] !== null) ? $stat['link_hits'] : 0;
+                $item['links_clicked'] = [];
+                if ($stat['link_url'] !== null) {
+                    $item['links_clicked'][] = $stat['link_url'];
+                }
+                /*if (isset($stat['campaign_id'])) {
+                    $item['campaign_id']   = $stat['campaign_id'];
+                    $item['campaign_name'] = $stat['campaign_name'];
+                }*/
+                if (isset($stat['segment_id'])) {
+                    $item['segment_id']   = $stat['segment_id'];
+                    $item['segment_name'] = $stat['segment_name'];
+                }
+                if (isset($stat['company_id'])) {
+                    $item['company_id']   = $stat['company_id'];
+                    $item['company_name'] = $stat['company_name'];
+                }
+                $data[$statId] = $item;
+            } else {
+                if ($stat['link_hits'] !== null) {
+                    $data[$statId]['click'] += $stat['link_hits'];
+                }
+                if ($stat['link_url'] !== null && !in_array($stat['link_url'], $data[$statId]['links_clicked'])) {
+                    $data[$statId]['links_clicked'][] = $stat['link_url'];
+                }
+            }
+        }
+
+        return $data;
+    }
+
+    /**
+     * @param           $limit
+     * @param \DateTime $dateFrom
+     * @param \DateTime $dateTo
+     * @param array     $options
+     * @param int|null  $companyId
+     * @param int|null  $campaignId
+     * @param int|null  $segmentId
+     *
+     * @return array
+     */
+    public function getMostHitEmailRedirects($limit, \DateTime $dateFrom, \DateTime $dateTo, $options = [], $companyId = null, $campaignId = null, $segmentId = null)
+    {
+        $createdByUserId = null;
+        if (!empty($options['canViewOthers'])) {
+            $createdByUserId = $this->userHelper->getUser()->getId();
+        }
+
+        $redirects = $this->redirectRepository->getMostHitEmailRedirects($limit, $dateFrom, $dateTo, $createdByUserId, $companyId, $campaignId, $segmentId);
+        $data      = [];
+        foreach ($redirects as $redirect) {
+            $data[] = [
+                'url'         => $redirect['url'],
+                'unique_hits' => $redirect['unique_hits'],
+                'hits'        => $redirect['hits'],
+                'email_id'    => $redirect['email_id'],
+                'email_name'  => $redirect['email_name'],
+            ];
+        }
```

Comment:
I can't figure out what this foreach do. Isn't `$data` and `$redirects` exactly the same in the end?

Reply:
I actually forgot select 'id' so it wasn't same. Removed now so it is same. However this is best practice because this mapping means changing context from entity to later used data structure. Same just by accident so if we would like to rename something we would have to add it.

Reply:
this is useful just to order the keys within the array and remove unused keys.